src/engine/BoundManager.cpp

```diff
@@ -0,0 +1,172 @@
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+
+**/
+
+#include "BoundManager.h"
+#include "Debug.h"
+#include "FloatUtils.h"
+#include "InfeasibleQueryException.h"
+#include "Tableau.h"
+#include "Tightening.h"
+
+using namespace CVC4::context;
+
+BoundManager::BoundManager( Context &context )
+    : _context( context )
+    , _size( 0 )
+    , _tableau( nullptr )
+{
+};
+
+BoundManager::~BoundManager()
+{
+    for ( unsigned i = 0; i < _size; ++i)
+    {
+        _lowerBounds[i]->deleteSelf();
+        _upperBounds[i]->deleteSelf();
+        _tightenedLower[i]->deleteSelf();
+        _tightenedUpper[i]->deleteSelf();
+    }
+};
+
+void BoundManager::initialize( unsigned numberOfVariables )
+{
+    ASSERT( 0 == _size );
+
+    for ( unsigned i = 0; i < numberOfVariables; ++i )
+        registerNewVariable();
+
+    ASSERT( numberOfVariables == _size );
+}
+
+unsigned BoundManager::registerNewVariable()
+{
+    ASSERT( _lowerBounds.size() == _size );
+    ASSERT( _upperBounds.size() == _size );
+    ASSERT( _tightenedLower.size() == _size );
+    ASSERT( _tightenedUpper.size() == _size );
+
+    unsigned newVar = _size++;
+
+    _lowerBounds.append( new ( true ) CDO<double>( &_context ) );
+    _upperBounds.append( new ( true ) CDO<double>( &_context ) );
+    _tightenedLower.append( new ( true ) CDO<bool>( &_context ) );
+    _tightenedUpper.append( new ( true ) CDO<bool>( &_context ) );
+
+    *_lowerBounds[newVar] = FloatUtils::negativeInfinity();
+    *_upperBounds[newVar] = FloatUtils::infinity();
+    *_tightenedLower[newVar] = false;
+    *_tightenedUpper[newVar] = false;
+
+    return newVar;
+}
+
+unsigned BoundManager::getNumberOfVariables()
+{
+    return _size;
+}
+
+bool BoundManager::tightenLowerBound( unsigned variable, double value )
+{
+    bool tightened = setLowerBound( variable, value );
+    if ( tightened &&  nullptr != _tableau )
+        _tableau->ensureNonBasicVariableGTLB( variable, value );
+    return tightened;
+}
+
+bool BoundManager::tightenUpperBound( unsigned variable, double value )
+{
+    bool tightened = setUpperBound( variable, value );
+    if ( tightened && nullptr != _tableau )
+        _tableau->ensureNonBasicVariableLTUB( variable, value );
+    return tightened;
+}
+
+bool BoundManager::setLowerBound( unsigned variable, double value )
+{
+    ASSERT( variable < _size );
+    if ( value > getLowerBound( variable ) )
+    {
+        *_lowerBounds[variable] = value;
+        *_tightenedLower[variable] = true;
+        if ( !consistentBounds( variable ) )
+            throw InfeasibleQueryException();
+        return true;
+    }
+    return false;
+}
+
+bool BoundManager::setUpperBound( unsigned variable, double value )
```

Comment:
This is not a problem, but an inconsistency with Tableau:
Traditionally, tighten() checks whether the new bound is tighter than the original, and only tightens in that case; whereas set just sets the bound, regardless of what was there before.

Set was typically used for initialization or for backtracking, when the bounds were actually made looser. 

I'm fine with keeping it as is, just wanted to point this out.

Reply:
Thanks for pointing this out, it is good to keep in mind. 

The difference I was going for is that set local to BoundManager, whereas tighten is the polite method to be used by various sources of tighter bounds. 
 
It would be more efficient to check whether the bound is tighter before sending it to the bound manager anyway. However, I am skeptical of simply over-writing bounds though, because any bounds we set must progress the search. Backtracking is lazy and does not invoke any of these methods. Overall, It should be OK to move the check to tighten*, but I need to test it on the branch with all context-dependent stuff to ensure it does not break anything.

Reply:
So upon reflecting, I think the implementation in the PR is more in line with how an SMT solver would handle a conflict. We arrive to a conflicted state from which conflict analysis repairs into a search state. Since we will be adding conflict analysis in the future, I will leave it like this.